### ISSUE TITLE
Consider moving our storage from AWS to Digital Alliance

### ISSUE THREAD

**Issue by jcohenadad** [spine-generic/data-multi-subject#162]
The [spine-generic](https://github.com/spine-generic/data-multi-subject) dataset is being increasingly downloaded, which comes at a cost. For example, the cost for 2023 was 478$, which is not negligible (cost per month below):
~~~
>>> 34+33+26+24+37+37+40+90+32+24+45+56
478
~~~

I'm wondering how feasible/difficult it would be to move the git-annex server to a Digital Alliance cloud?

**Comment by namgo** [comment]
I can't remember off the top of my head what Digital Alliance's policy is for this but I know I'd talked to Nick about it before. My memory is that they're happy to supply the bandwidth but I could be mistaken.

I'll submit a ticket about it now.

**Comment by mguaypaq** [comment]
It should be possible, but I'll just point out a technical difference:
* For our internal dataset server, and for spineimage.ca, we're running NeuroGitea, which has both the repository metadata (file names, directories, commits, etc.) _and_ the large data files (the actual .nii.gz files).
* For spine-generic, we still want to host the repository metadata on Github. It's just the large data files that we want to host on the Digital Alliance cloud, instead of on Amazon S3. 

So, it's not exactly the same interface/setup. But, I'm sure there exists open source software which can present the same interface as S3. Or, we can also look at the other types of [special remotes](https://git-annex.branchable.com/special_remotes/) that git-annex supports.

**Comment by jcohenadad** [comment]
Thank you for the clarification @mguaypaq. Would it make sense to move spine-generic to a NeuroGitea server?

**Comment by mguaypaq** [comment]
I think Github still gives us a lot of value:
- All the existing links keep working.
- All the issues/pull requests are there.
- We don't have to manage user accounts for every collaborator.

Probably we just want to change the storage backend for the large data files, that should be a much smaller change from the point of view of our users. It should be very doable.

**Comment by mguaypaq** [comment]
In particular, it looks like the Digital Alliance already has a service that's compatible with S3, so maybe this can be easy:
https://docs.alliancecan.ca/wiki/Arbutus_object_storage

**Comment by namgo** [comment]
I heard back from Digital Alliance on Arbutus (way quicker than I expected... 30 minute turnaround time!).

There's no specific policy about using arbutus for public datasets but the bandwidth they can provide is pretty limited - unless it's primarily our own users who are downloading the dataset? They have fast uplinks to Canadian university networks, but relatively slow uplinks to everyone else from the sounds of things?

Do we know how many external people are using the dataset?

**Comment by jcohenadad** [comment]
> Do we know how many external people are using the dataset?

i'd say ~10 ppl/month? but uplinks doesn't matter too much. What matters is downlinks. And as long as it is not insanely slow (which is not), we should be fine

**Comment by namgo** [comment]
I mean uplink in this case would mean their upload speed to external networks which would affect download speed but:

> Well the CANARIE research backbone that connects all of Canada together has some pretty good interconnections with Internet2 and GEANT at the very least, so that should cover the USA and most of Europe.

I think that fits our needs, very cool!

@mguaypaq do you have a vision for how you might switch backends in git-annex? I imagine we can just import the dataset from s3 to arbutus without much trouble but I'm not sure about how this works with git-annex.

**Comment by mguaypaq** [comment]
Just like git, git-annex supports having multiple remotes. So, I imagine we would:
1. Figure out the right configuration and permissions for Arbutus object storage. (Probably you @namgo? Although maybe I can help since I have admin access to the existing Amazon S3 stuff.)
2. Configure Arbutus as a new special remote for git-annex, alongside the existing Amazon special remote. (Probably me, since I'm most familiar with git-annex.)
3. Copy over the files through git-annex, which will double as a test that new files _can_ be uploaded by people with write access.
4. Test that a new clone can get the files from the Arbutus special remote.
5. Deconfigure the Amazon special remote from git-annex (but keep the files in place for a little while).
6. Once everything works, delete the Amazon buckets.

So, a nice gradual transition, with plenty of opportunity to roll back if there are problems.

**Comment by mguaypaq** [comment]
That's possible! We know that the QC report generator only copies _missing_ asset files (js, css, etc.), and assumes that the existing ones are all up-to-date:

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/1af24e30cdae0b465f6abd33c5633b6bcb71de39/spinalcordtoolbox/reports/qc.py#L624-L625

**Comment by namgo** [comment]
@nullnik-0  is becoming our resident expert on ComputeCanada already! I'd be down to work alongside her since I have admin perms on our CC projects. I'll loop all three of us into a slack convo and we could talk about permissions.

**Comment by mguaypaq** [comment]
Preliminary tests seem to work! We should be able to migrate to Arbutus object storage fairly quickly and reduce our Amazon bandwidth costs.

Steps:

* Following [the Alliance docs for Arbutus object storage](https://docs.alliancecan.ca/wiki/Arbutus_object_storage), I downloaded and sourced my OpenStack RC from [the API access dashboard](https://arbutus.cloud.computecanada.ca/project/api_access/). Then I ran

  ```sh
  openstack ec2 credentials create
  ```
  and saved the resulting access key and secret key in my password manager.

* In a fresh clone of [spine-generic/data-single-subject](https://github.com/spine-generic/data-single-subject), I created two special remotes:

  * `arbutus-read`: this one is world-readable and auto-enabled.

  * `arbutus-write`: this one has to be enabled manually, and can be used by people with ec2 credentials (like me from the previous point) to upload image files. Git-annex knows that it refers to the same bucket as `arbutus-read`.

  ```sh
  read -r AWS_ACCESS_KEY_ID AWS_SECRET_ACCESS_KEY
  # copy-paste the access key and secret key, separated by a space, then press enter
  export AWS_ACCESS_KEY_ID AWS_SECRET_ACCESS_KEY
  env | grep AWS_

  git annex initremote arbutus-read type=S3 \
    autoenable=true \
    bucket=def-jcohen-data-single-subject \
    datacenter=CA \
    encryption=none \
    host=object-arbutus.cloud.computecanada.ca \
    port=443 \
    protocol=https \
    public=yes \
    publicurl=https://object-arbutus.cloud.computecanada.ca/def-jcohen-data-single-subject/ \
    requeststyle=path

  git annex initremote --sameas=arbutus-read arbutus-write type=S3 \
    bucket=def-jcohen-data-single-subject \
    datacenter=CA \
    host=object-arbutus.cloud.computecanada.ca \
    port=443 \
    protocol=https \
    public=no \
    requeststyle=path
  ```

* Get some data files and push them to Arbutus:

  ```sh
  git annex get sub-douglas
  git annex copy --to arbutus-write sub-douglas
  ```

* In a clone-of-my-clone, try to get the files, without having ec2 credentials:

  ```sh
  unset AWS_ACCESS_KEY_ID AWS_SECRET_ACCESS_KEY
  git annex get --from arbutus-read sub-douglas
  ```

I'm out of time for this week, but next week I'll try to migrate both `data-single-subject` and `data-multi-subject` to Arbutus.

**Comment by bpinsard** [comment]
Stumbling on that convo... 
You do not necessarily need to decommission the AWS storage as you can manage which special-remote gets chosen by default (while keeping others as a fallback) by setting a `cost` value to the special remote when initializing it (or adding it with `enableremote` afterward ). 

Also, I don't think you need to set 2 special remotes for read/write respectively, the first one should work for both (write with credentials only).

**Comment by jcohenadad** [comment]
thank you for your insights @bpinsard!

a fallback is definitely useful, but don't we already have a backup of spine-generic @namgo ?

**Comment by mguaypaq** [comment]
@bpinsard have you gotten the single read/write remote to work in the past? I remember trying (in the past year) to use a single special remote for both, but couldn't get it to work. Possibly something to do with [this interaction between the config settings](http://git-annex.branchable.com/special_remotes/S3/#comment-fcfba0021592de4c1425d3bf3c9563d3).

**Comment by bpinsard** [comment]
We do not use that setup in production (only authenticated access), but I just tested it with a minio s3 server (so not the same as digital alliance). 
There is one caveat though (which I think occurs with the split setup too), is if someone has permanently setup AWS keys in their environment for another server/usage.
In that case, it overrides the anonymous access and can cause 403 errors because git-annex sends the credentials (it might depend on how each server deals with policies and anonymous access). 

I think a good way to avoid that, is to set the S3 remote for the read/write data management only, not autoenabled, and then add a httpalso sameas remote, crafting the https url depending on the server, bucket-name and requeststyle.

```
git annex initremote https_download --sameas=s3_remote_name autoenable=true type=httpalso url=https://s3.unf-montreal.ca/test.publicurl/ cost=50 
```

This can save a lot of user-support headaches.

**Comment by mguaypaq** [comment]
Oh! I didn't know about the httpalso remote type, that makes a lot of sense. It's still two remotes with a sameas, but probably with fewer corner cases.

**Comment by namgo** [comment]
> a fallback is definitely useful, but don't we already have a backup of spine-generic @namgo ?

We have a backup of what's on gitea (and what was on gitolite) however those I understand to be git-annex archives rather than the datasets themselves.

**Comment by namgo** [comment]
Whoops! I misunderstood - we don't have spine-generic backed up on restic - mathieu helped me remember that that one's on github.

**Comment by jcohenadad** [comment]
then we should probably create a backup, no?

**Comment by namgo** [comment]
Good point. I made a ticket for getting this put in restic (with some questions for Mathieu) - should be pretty straightforward.

**Comment by mguaypaq** [comment]
Ok, I'm going to do the same migration from AWS to AllianceCan as in https://github.com/spine-generic/data-single-subject/issues/35 for `data-multi-subject`. We will still have AWS accessible for now in case of problems, but git-annex will be configured to prefer AllianceCan.

Once this is done, I'll update the documentation in https://github.com/spine-generic/spine-generic/wiki/git-annex.

**Comment by bpinsard** [comment]
YAY, less money for Bezos :D 
Have you considered having a copy of the annexed files at CONP?
And/or an archival at https://borealisdata.ca/ as https://github.com/datalad/datalad-dataverse is supposed to support that.
My understanding with dataverse is that it can also archive the git part of the dataset, as not to rely on github in case the US continues to go down in flames. Though you have your own gitea IIRC.

**Comment by jcohenadad** [comment]
> Have you considered having a copy of the annexed files at CONP?
And/or an archival at https://borealisdata.ca/ as https://github.com/datalad/datalad-dataverse is supposed to support that.

great idea

**Comment by mguaypaq** [comment]
Hmm, it looks like the Compute Canada remote is slower than the Amazon one:
- 98 minutes to `git annex get` from Compute Canada.
- 110 minutes (!) when I tried to `git annex get` from Compute Canada again while _skipping_ the `derivatives/` folder.
- 58 minutes to `git annex get` from Amazon.

I think we can live with this slowdown. The only major problem would be if we can't run the automated [Dataset Validator](https://github.com/spine-generic/data-multi-subject/blob/8fafd833c63483c1d5f8aa3936b246029bb42fa9/.github/workflows/validator.yml) within the time limits imposed by Github Actions.

But even in that case, maybe we can still cover most of our bases by:
- For pull requests, skip the very long download, and only run `bids-validator --ignoreNiftiHeaders`.
  - Maybe also `git annex get` only the files that are touched by the PR, as a quick check that they are correctly uploaded?
- Add a new scheduled (maybe daily or weekly) workflow, which does the more expensive checking periodically, instead of being triggered on every pull request.

For now, I'm going to "merge" the migration anyway, since it doesn't remove the Amazon remote completely. I can deal with the automated workflows later if need be.

**Comment by mguaypaq** [comment]
BTW, here's the new special remote settings for `data-multi-subject`:

```
33ef5eee-44fa-4467-98b0-bfbfbb8cb758
  name=do-not-use-1
    timestamp=1672953055.042966s

5a5447a8-a9b8-49bc-8276-01a62632b502
    autoenable=false
    bucket=data-multi-subject---spine-generic---neuropoly 
    cost=300 
    datacenter=ca-central-1 
    encryption=none 
    host=s3.amazonaws.com 
  name=amazon-private 
    port=443 
    signature=v4 
    storageclass=STANDARD 
    type=S3 
    timestamp=1749579528s

5e339d08-5eff-4912-b262-d0c2b1342fcd
    autoenable=true
    cost=300
  sameas-name=amazon-public
    sameas-uuid=5a5447a8-a9b8-49bc-8276-01a62632b502
    type=httpalso 
    url=https://data-multi-subject---spine-generic---neuropoly.s3.ca-central-1.amazonaws.com 
    timestamp=1749579674s

8633c5a1-4415-4437-8f91-5fef69695c8a 
    autoenable=true
  sameas-name=computecanada-public
    sameas-uuid=afd7e696-7b3a-4c7e-9dd1-4dfa87cdbd31
    type=httpalso
    url=https://object-arbutus.cloud.computecanada.ca/def-jcohen-data-multi-subject
    timestamp=1749580641s

8ffcc392-0777-44d4-bec4-42a7aeee018b
  name=do-not-use-2
    timestamp=1672953143.199932s

9af40aaa-e68c-44a6-9e75-de323b371de4 
  name=do-not-use-3 
    timestamp=1672952808.532591s

afd7e696-7b3a-4c7e-9dd1-4dfa87cdbd31 
    autoenable=false
    bucket=def-jcohen-data-multi-subject 
    datacenter=CA
    encryption=none 
    host=object-arbutus.cloud.computecanada.ca 
  name=computecanada-private 
    port=443 
    requeststyle=path
    storageclass=STANDARD
    type=S3 
    timestamp=1749580611s
```

(The `do-not-use` remotes are leftovers from previous experiments, and `git-annex` tries to revive them at every opportunity, so the only sensible option is to keep them around with such a name.)

**Comment by bpinsard** [comment]
If you want faster tests that data are present in the remote `git-annex fsck --fast --from remotename`  will just HEAD the file from the remote, it won't check the hash though, but it at least test connectivity and file presence (and authentication if set-up). That's what we do for cneuromod datasets otherwise downloading terabytes of data to GH actions would take way too long.

**Comment by mguaypaq** [comment]
Good point, thanks. We probably still want periodic checks of the data itself (for example, `bids-validator` does open the nifti files to check some of the header values), but a fast file presence test for any new files should be good enough for pull requests.
